ocaml/xapi/xapi_pool.ml

```diff
@@ -807,6 +807,33 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
               (pool_joining_host_ca_certificates_conflict, !conflicting_names)
           )
   in
+  let assert_no_host_pending_mandatory_guidance () =
+    (* Assert that there is no host pending mandatory guidance on the joining
+       host or the pool coordinator.
+    *)
+    Repository_helpers.assert_no_host_pending_mandatory_guidance ~__context
+      ~host:(Helpers.get_localhost ~__context) ;
+    let master = get_master ~rpc ~session_id in
+    let master_pending_mandatory_guidances =
```

Comment:
This only checks pending guidances on the coordinator. Shouldn't the guidances be checked for all hosts instead?

Reply:
I think we only need to check the guidances on the coordinator as it is the "joining host" and coordinator whose "software_version", "api versiion" and "db schema" are checked in the `pre_join_checks`. Outstanding mandatory guidances in other hosts will not affect pool.join.

Reply:
But I think outstanding guidances on other pool members will not affect pool join, and you can apply them before or after a new host joins the pool, and with one more host in the pool, it might be helpful for some operation like VM migration or host evacuation.

Reply:
True if you look at in isolation. But you are dragging along an undefined situation. It's called mandatory for a reason.

Reply:
Yes, agree, outstanding mandatory guidance on any host results in an undefined situation.
But I think it is not only "pool.join" which should avoid operating in such situation, it should be avoided for any pool operation. While we will not check for it in any XAPI API call, we will only check it if the outstanding mandatory guidance will have a direct consequence for the API call, like "host.apply_updates" and "pool.join".

Reply:
In current design, I think it is XenCenter's responsibility to guide user to apply all outstanding mandatory guidance after a pool upgrade.